### PR TITLE
- added inital proof of concept of roslyn integration

### PR DESCRIPTION
**Pull request by asd-and-Rizzo**

### CODE REVIEW

File: ExpressionToCode.Roslyn/ExpressionsConverter.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        
+        
+        public static SyntaxTree ToSyntaxTree(Expression e) {
+            //TODO: slow but working hack, fix it my modifying ExpressionToCode visitor to allow emit Roslyn SyntaxTree
+            return CSharpSyntaxTree.ParseText(ExpressionToCodeLib.ExpressionToCode.ToCode(e));
+        }
+        
+        
+        public static CSharpCompilation ToCompilationUnit<T>(Expression<Func<T>> e) { return ToCompilationUnit((Expression)e); }
+
```
Comment (EamonNerbonne) [inline ExpressionToCode.Roslyn/ExpressionsConverter.cs]: What does this do that normal expression tree compilation cannot?

**Comment by asd-and-Rizzo** [comment]
Roslyn may wrap expression into method for newly created class. Expression cannot compile new class or new method. One possible use case when DI containers based on Expressions embrace code generation as part of resolution (interception, AOP, patterns automation, container self compilation into class with less of container for performance and other reasons).

As I said here https://github.com/EamonNerbonne/ExpressionToCode/issues/31 Roslyn may be searched for feature to jump from Expression to CSharpCompilation to avoid performance overheads.


If extend question to syntax tree - then cross generation to VB, custom code formatting.

**Comment by EamonNerbonne** [comment]
`LambdaExpression` has a [`CompileToMethod`](https://msdn.microsoft.com/en-us/library/dd728258%28v=vs.110%29.aspx) - and there's also the delegate-based compilation which ExpressionToCode currently uses.

**Comment by asd-and-Rizzo** [comment]
Expression compile into IL body, not method declaration.  Hence C# oriented people have to work on IL level to define  class and method, and Expression  provides only IL. With Roslyn all may go AST language oriented way. Differences in details of usability, analysis and debug methods, and outputs not only of IL.

**Comment by asd-and-Rizzo** [comment]
Here is first use case out of https://bitbucket.org/Dzmitry_Lahoda/dryioc/branch/edge to write custom dispatcher for Roslyn instead of ParseHack:


```csharp
       private static MethodDeclarationSyntax createRootFactoryMethod(ResolutionRootNode root)
        {
            //TODO: fix this rude hack which will fail because of ambiguilty of parsing code snipped with direct conversion of NewExpression into ObjectCreationExpressionSyntax
            var parsed = CSharpSyntaxTree.ParseText(ExpressionToCode.ToCode(root.Factory), new CSharpParseOptions(LanguageVersion.CSharp5, DocumentationMode.None, SourceCodeKind.Script)).GetRoot();
            var creation = ((ExpressionStatementSyntax)((GlobalStatementSyntax)parsed.ChildNodes().ToArray()[0]).Statement).Expression as ObjectCreationExpressionSyntax;
            return SyntaxFactory.MethodDeclaration(
                                  SyntaxFactory.IdentifierName(
                                     root.ReturnType.ToString().Replace("+", ".")),
                                  SyntaxFactory.Identifier(
                                      "Get" + root.ReturnType.Name))
                              .WithModifiers(
                                  SyntaxFactory.TokenList(
                                      new[]{
                                        SyntaxFactory.Token(
                                            SyntaxKind.PublicKeyword),
                                        SyntaxFactory.Token(
                                            SyntaxKind.StaticKeyword)}))
                              .WithBody(
                                  SyntaxFactory.Block(
                                      SyntaxFactory.SingletonList<StatementSyntax>(
                                          SyntaxFactory.ReturnStatement(creation))));
        }
```

**Pull request merged**